### ISSUE TITLE
Static linking would benefit Android usage

### ISSUE THREAD

**Issue by staltz** [holepunchto/sodium-native#111]
I'm building https://github.com/staltz/sodium-native-nodejs-mobile as a fork of this module, but just as a quick hack for now, to get things working (it's much easier if I don't have to generalize). Eventually I intend to send a PR here to support Android.

I hit an issue https://github.com/JaneaSystems/nodejs-mobile-module-compat/issues/5 related to `dlopen`ing libsodium.so. The shared library .so is not loaded by the mobile app (because the build process doesn't manage to configure the mobile app to load it). 

Similar to #100, it would be easier if libsodium were be a static library. This would require (I think) there to be a gyp config in the libsodium submodule.

**Comment by emilbayes** [comment]
Can you try to debug what path it is trying to load? For linux we have to set the rpath like this: https://github.com/sodium-friends/sodium-native/blob/master/binding.gyp#L44

On macOS we have to set it like this: https://github.com/sodium-friends/sodium-native/blob/master/postinstall.js#L56-L58

There's unfortunately a lot of tacit knowledge in how to make the dynamic loader happy. We can also chat on IRC in #sodium-friends on freenode if that makes it easier :)

**Comment by emilbayes** [comment]
Maybe the magic variables are named differently on android? https://android.googlesource.com/platform/bionic/+/master/android-changes-for-ndk-developers.md#dt_runpath-support-available-in-api-level-24

**Comment by emilbayes** [comment]
Can you somehow set `LD_DEBUG` on Android and have it write to logcat? This should make the dynamic loader / linker write out what it is trying to do

**Comment by staltz** [comment]
Thanks for the tips!

Sanity check: I put the `.so` files in `android/app/src/main/jniLibs/${arch}/` and my app has a thing called SoLoader (from React Native) which knows how to look those files up and load them. And now it's working in runtime. But I don't want to do that, it's too manual.

Next, I'll try `runpath` instead of `rpath`.

**Comment by emilbayes** [comment]
Also, what ours say is that the libsodium.so should be loaded from the same folder as the sodium.node shared library. So it all depends on where that one ends up

**Comment by staltz** [comment]
I did this in bindings.gyp:

```
'link_settings': {
  'libraries': [
    '-Wl,--enable-new-dtags',
    '-Wl,-rpath=\\$$ORIGIN'
  ]
},
```

And the ELF header for `sodium.node` now looks like this:

`$ANDROID_NDK/toolchains/aarch64-linux-android-4.9/prebuilt/linux-x86_64/bin/aarch64-linux-android-readelf -d build/Release/sodium.node`

```
Dynamic section at offset 0x376e8 contains 35 entries:
  Tag        Type                         Name/Value
 0x00000003 (PLTGOT)                     0x38a7c
 0x00000002 (PLTRELSZ)                   2800 (bytes)
 0x00000017 (JMPREL)                     0xa670
 0x00000014 (PLTREL)                     REL
 0x00000011 (REL)                        0x9f08
 0x00000012 (RELSZ)                      1896 (bytes)
 0x00000013 (RELENT)                     8 (bytes)
 0x6ffffffa (RELCOUNT)                   6
 0x00000006 (SYMTAB)                     0x1cc
 0x0000000b (SYMENT)                     16 (bytes)
 0x00000005 (STRTAB)                     0x228c
 0x0000000a (STRSZ)                      24532 (bytes)
 0x6ffffef5 (GNU_HASH)                   0x8260
 0x00000004 (HASH)                       0x8a58
 0x00000001 (NEEDED)                     Shared library: [libsodium.so]
 0x00000001 (NEEDED)                     Shared library: [liblog.so]
 0x00000001 (NEEDED)                     Shared library: [libnode.so]
 0x00000001 (NEEDED)                     Shared library: [libc++_shared.so]
 0x00000001 (NEEDED)                     Shared library: [libm.so]
 0x00000001 (NEEDED)                     Shared library: [libdl.so]
 0x00000001 (NEEDED)                     Shared library: [libc.so]
 0x0000000e (SONAME)                     Library soname: [sodium.node]
 0x0000001a (FINI_ARRAY)                 0x38520
 0x0000001c (FINI_ARRAYSZ)               8 (bytes)
 0x00000019 (INIT_ARRAY)                 0x386e4
 0x0000001b (INIT_ARRAYSZ)               4 (bytes)
 0x0000001d (RUNPATH)                    Library runpath: [$ORIGIN]
 0x0000001e (FLAGS)                      BIND_NOW
 0x6ffffffb (FLAGS_1)                    Flags: NOW
 0x6ffffff0 (VERSYM)                     0x9ab4
 0x6ffffffc (VERDEF)                     0x9ecc
 0x6ffffffd (VERDEFNUM)                  1
 0x6ffffffe (VERNEED)                    0x9ee8
 0x6fffffff (VERNEEDNUM)                 1
 0x00000000 (NULL)                       0x0
```

Notice

```
0x0000001d (RUNPATH)                    Library runpath: [$ORIGIN]
```

So it looks correct. But even if `libsodium.so` is in `build/Release/`, after the build of the Android app, it doesn't seem to work out, but I think it might be more related to nodejs-mobile and/or Android gradle.

**Comment by emilbayes** [comment]
Hmm this tells us that it needs libsodium.so but not where it tries to load from. Is there something like ldd that can be run on the device to see what it resolves to?

**Comment by RangerMauve** [comment]
According to cel on SSB, it should be possible to make this work on Android with the following patch:

https://viewer.scuttlebot.io/%25Ik60jkfEOEBvyxz5nuhNQ7vNu10Rx0qXlEt%2B3hOWbBo%3D.sha256

**Comment by kasperisager** [comment]
This should be solved by #174 and #175.
